File: ios/Classes/lib/core/managers/CancellationManager.swift
//
//  CancellationManager.swift
//  awesome_notifications
//
//  Created by Rafael Setragni on 06/12/21.
//

import Foundation

public class CancellationManager {
    
    public static let TAG: String = "CancellationManager"
    
    // ************** SINGLETON PATTERN ***********************
    
    static var instance:CancellationManager?
    public static var shared:CancellationManager {
        get {
            CancellationManager.instance =
                CancellationManager.instance ?? CancellationManager()
            return CancellationManager.instance!
        }
    }
    private init(){}
    
    // ********************************************************
        
    public func dismissNotification(byId id:Int) -> Bool {
        return StatusBarManager
                    .shared
                    .dismissNotification(id: id)
    }
    
    public func dismissNotifications(byChannelKey channelKey: String) -> Bool {
        return StatusBarManager
                    .shared
                    .dismissNotificationsByChannelKey(channelKey: channelKey)
    }
    
    public func dismissNotifications(byGroupKey groupKey: String) -> Bool {
        return StatusBarManager
                    .shared
                    .dismissNotificationsByGroupKey(groupKey: groupKey)
    }
    
    public func dismissAllNotifications() -> Bool {
        return StatusBarManager
                    .shared
                    .dismissAllNotifications()
    }
    
    public func cancelSchedule(byId id:Int) -> Bool {
        return
            ScheduleManager.cancelScheduled(id: id) &&
            _cancelNativeScheduledNotification(id: id)
    }
    
    public func cancelSchedules(byChannelKey channelKey: String) -> Bool {
        return
            _cancelNativeSchedulesByChannelKey(channelKey: channelKey)
    }
    
    public func cancelSchedules(byGroupKey groupKey: String) -> Bool {
        return
            _cancelNativeSchedulesByGroupKey(groupKey: groupKey)
    }
    
    public func cancelAllSchedules() -> Bool {
        return
            ScheduleManager.cancelAllSchedules() &&
            _cancellAllNativeScheduledNotifications()
    }
    
    public func cancelNotification(byId id:Int) -> Bool {
        return
            dismissNotification(byId: id) &&
            cancelSchedule(byId: id)
    }
    
    public func cancelNotifications(byChannelKey channelKey: String) -> Bool {
        return
            dismissNotifications(byChannelKey: channelKey) &&
            cancelSchedules(byChannelKey: channelKey)
    }
    
    public func cancelNotifications(byGroupKey groupKey: String) -> Bool {
        return
            dismissNotifications(byGroupKey: groupKey) &&
            cancelSchedules(byGroupKey: groupKey)
    }

    public func cancelAllNotifications() -> Bool {
        return
            dismissAllNotifications() &&
            cancelAllSchedules()
    }
        
    private func _cancelNativeSchedulesByChannelKey(channelKey: String) -> Bool {
        
        let center = UNUserNotificationCenter.current()
        
        center.getPendingNotificationRequests(completionHandler: { (notificationRequest) in
            for notification in notificationRequest {
                if channelKey == notification.content.userInfo[Definitions.NOTIFICATION_CHANNEL_KEY] as? String {
                    if notification.content.userInfo[Definitions.NOTIFICATION_ID] != nil {
                        
                        if let id:String = notification.content.userInfo[Definitions.NOTIFICATION_ID] as? String {
                            center.removePendingNotificationRequests(withIdentifiers: [id])
                            let _ = ScheduleManager.cancelScheduled(id: Int(id)!)
                        }
                        else {
                            if let id:Int64 = notification.content.userInfo[Definitions.NOTIFICATION_ID] as? Int64 {
                                center.removePendingNotificationRequests(withIdentifiers: [String(id)])
                                let _ = ScheduleManager.cancelScheduled(id: Int(id))
                            }
                        }
                    }
                }
            }
        })
        
        return true
    }
    
    private func _cancelNativeSchedulesByGroupKey(groupKey: String) -> Bool {

        let center = UNUserNotificationCenter.current()
        
        center.getPendingNotificationRequests(completionHandler: { (notificationRequest) in
            for notification in notificationRequest {
                if groupKey == notification.content.userInfo[Definitions.NOTIFICATION_GROUP_KEY] as? String {
                    if notification.content.userInfo[Definitions.NOTIFICATION_ID] != nil {
                        
                        if let id:String = notification.content.userInfo[Definitions.NOTIFICATION_ID] as? String {
                            center.removePendingNotificationRequests(withIdentifiers: [id])
                            let _ = ScheduleManager.cancelScheduled(id: Int(id)!)
                        }
                        else {
                            if let id:Int64 = notification.content.userInfo[Definitions.NOTIFICATION_ID] as? Int64 {
                                center.removePendingNotificationRequests(withIdentifiers: [String(id)])
                                let _ = ScheduleManager.cancelScheduled(id: Int(id))
                            }
                        }
                    }
                }
            }
        })
        
        return true
    }
    
    private func _cancelNativeScheduledNotification(id:Int) -> Bool {
        let referenceKey:String = String(id)
            
        let center = UNUserNotificationCenter.current()
        center.removePendingNotificationRequests(withIdentifiers: [referenceKey])
        
        return true
    }
    
    private func _cancelNativeNotification(id:Int) -> Bool {
        let referenceKey:String = String(id)
            
        let center = UNUserNotificationCenter.current()
        center.removeDeliveredNotifications(withIdentifiers: [referenceKey])
        center.removePendingNotificationRequests(withIdentifiers: [referenceKey])
        
        return true
    }
    
    private func _cancellAllNativeScheduledNotifications() -> Bool {
            
        let center = UNUserNotificationCenter.current()
        center.removeAllPendingNotificationRequests()
        
        return true
    }
    
    public func _cancellAllNativeNotifications() -> Bool {
            
        let center = UNUserNotificationCenter.current()
        center.removeAllDeliveredNotifications()
        center.removeAllPendingNotificationRequests()
        
        return true
    }
    
}

File: example/android/awn_core/src/main/java/me/carda/awesome_notifications/core/managers/CancellationManager.java
package me.carda.awesome_notifications.core.managers;

import android.content.Context;

import androidx.annotation.NonNull;

import me.carda.awesome_notifications.core.exceptions.AwesomeNotificationsException;
import me.carda.awesome_notifications.core.exceptions.ExceptionCode;
import me.carda.awesome_notifications.core.exceptions.ExceptionFactory;
import me.carda.awesome_notifications.core.threads.NotificationScheduler;
import me.carda.awesome_notifications.core.utils.StringUtils;

public class CancellationManager {

    private static final String TAG = "CancellationManager";

    private final StringUtils stringUtils;
    
    // ************** SINGLETON PATTERN ***********************

    protected static CancellationManager instance;

    protected CancellationManager(StringUtils stringUtils){
        this.stringUtils = stringUtils;
    }

    public static CancellationManager getInstance() {
        if (instance == null)
            instance = new CancellationManager(
                    StringUtils.getInstance());
        return instance;
    }

    // ********************************************************

    public boolean dismissNotification(@NonNull final Context context, Integer notificationId) throws AwesomeNotificationsException {

        if (notificationId == null || notificationId < 0)
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_INVALID_ARGUMENTS,
                            "Invalid notification id",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".notificationId");

        StatusBarManager
                .getInstance(context)
                .dismissNotification(context, notificationId);

        return true;
    }

    public boolean dismissNotificationsByChannelKey(@NonNull final Context context, @NonNull final String channelKey) throws AwesomeNotificationsException {

        if(stringUtils.isNullOrEmpty(channelKey))
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_INVALID_ARGUMENTS,
                            "Invalid channel key",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".channelKey");

        StatusBarManager
                .getInstance(context)
                .dismissNotificationsByChannelKey(context, channelKey);

        return true;
    }

    public boolean dismissNotificationsByGroupKey(@NonNull final Context context, @NonNull final String groupKey) throws AwesomeNotificationsException {

        if(stringUtils.isNullOrEmpty(groupKey))
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_INVALID_ARGUMENTS,
                            "Invalid group key",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".groupKey");

        StatusBarManager
                .getInstance(context)
                .dismissNotificationsByGroupKey(context, groupKey);

        return true;
    }

    public void dismissAllNotifications(@NonNull final Context context) throws AwesomeNotificationsException {
        StatusBarManager
                .getInstance(context)
                .dismissAllNotifications(context);
    }

    public boolean cancelSchedule(@NonNull final Context context, Integer notificationId) throws AwesomeNotificationsException {

        if (notificationId == null || notificationId < 0)
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_INVALID_ARGUMENTS,
                            "Invalid notification id",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".notificationId");

        NotificationScheduler.cancelScheduleById(context, notificationId);

        return true;
    }

    public boolean cancelSchedulesByChannelKey(@NonNull final Context context, @NonNull final String channelKey) throws AwesomeNotificationsException {

        if(stringUtils.isNullOrEmpty(channelKey))
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_INVALID_ARGUMENTS,
                            "Invalid channel key",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".channelKey");

        NotificationScheduler.cancelSchedulesByChannelKey(context, channelKey);

        return true;
    }

    public boolean cancelSchedulesByGroupKey(@NonNull final Context context, @NonNull final String groupKey) throws AwesomeNotificationsException {

        if(stringUtils.isNullOrEmpty(groupKey))
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_INVALID_ARGUMENTS,
                            "Invalid group key",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".groupKey");

        NotificationScheduler.cancelSchedulesByGroupKey(context, groupKey);

        return true;
    }

    public void cancelAllSchedules(@NonNull final Context context) throws AwesomeNotificationsException {
        NotificationScheduler.cancelAllSchedules(context);
    }

    public boolean cancelNotification(@NonNull final Context context, Integer notificationId) throws AwesomeNotificationsException {

        if (notificationId == null || notificationId < 0)
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_INVALID_ARGUMENTS,
                            "Invalid notification id",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".notificationId");

        cancelSchedule(context, notificationId);
        dismissNotification(context, notificationId);

        return true;
    }

    public boolean cancelNotificationsByChannelKey(@NonNull final Context context, @NonNull final String channelKey) throws AwesomeNotificationsException {

        if(stringUtils.isNullOrEmpty(channelKey))
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_INVALID_ARGUMENTS,
                            "Invalid channel key",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".channelKey");

        dismissNotificationsByChannelKey(context, channelKey);
        cancelSchedulesByChannelKey(context, channelKey);

        return true;
    }

    public boolean cancelNotificationsByGroupKey(@NonNull final Context context, @NonNull final String groupKey) throws AwesomeNotificationsException {

        if(stringUtils.isNullOrEmpty(groupKey))
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_INVALID_ARGUMENTS,
                            "Invalid group key",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".groupKey");

        dismissNotificationsByGroupKey(context, groupKey);
        cancelSchedulesByGroupKey(context, groupKey);

        return true;
    }


    public void cancelAllNotifications(@NonNull final Context context) throws AwesomeNotificationsException {

        dismissAllNotifications(context);
        cancelAllSchedules(context);
    }
}
